GitExtUtils/FileLogging/Logger.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace GitExtUtils.FileLogging
+{
+    /// <summary>
+    /// For unit testing
+    /// </summary>
+    public class LoggerStub : ILogger
+    {
+        public List<LogEntry> LogEntries = new List<LogEntry>();
+
+        public void WaitShutdown()
+        {
+        }
+
+        public void Log(LogSeverity severity, string message, Dictionary<string, string> extra = null)
+        {
+            LogEntries.Add(LogEntryCreator.Create(severity, message, extra));
+        }
+
+        public void Info(string message, Dictionary<string, string> extra = null)
+        {
+            LogEntries.Add(LogEntryCreator.Create(LogSeverity.Info, message, extra));
+        }
+
+        public void Error(string message, Dictionary<string, string> extra = null)
+        {
+            LogEntries.Add(LogEntryCreator.Create(LogSeverity.Error, message, extra));
+        }
+
+        public void Error(string message, Exception e, Dictionary<string, string> extra = null)
+        {
+            LogEntries.Add(LogEntryCreator.Create(LogSeverity.Error, message, e, extra));
+        }
+
+        public void LogGitCommandExecution(string message, string fileName, string arguments, string workingDir, int? processId,
+            bool isOnMainThread, TimeSpan? duration, int? exitCode, StackTrace callStack)
+        {
+            LogEntries.Add(LogEntryCreator.Create(message, fileName, arguments, workingDir, processId, isOnMainThread, duration,
+                exitCode, callStack));
+        }
+    }
+
+    public class Logger : ILogger
+    {
+        private readonly LogWriter _writer;
+
+        private readonly Task _writerTask;
+
+        public const string MetaTypeKeyName = "_MetaType";
+
+        private static ILogger _instance = new LoggerStub();
```

Comment:
Why do we need this?

Reply:
Its there for 2 reasons.

1. To ensure that we can execute code during unit test without fixing up all the unit tests
2. To enable unit tests to assert that logging has been taken place. In some areas logging may be so essential you want to ensure it through testing.

recall, the static logger is here as a migration path towards instance based logging. Since logging is/will permeate most of the code base, it will be a significant task to pass around logger instances all the places and fixing up all unit tests that execute logging code.

the static logger is a way to introduce logging and enable more people help gradually convert the code base.